interfaces/prompting/patterns/scan.go

```diff
@@ -0,0 +1,129 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patterns
+
+import (
+	"errors"
+	"io"
+	"strings"
+)
+
+type TokenType int
+
+const (
+	tokEOF TokenType = iota
+	tokError
+	tokText
+	tokBraceOpen
+	tokBraceClose
+	tokComma
+)
+
+func (t TokenType) String() string {
+	switch t {
+	case tokEOF:
+		return "end-of-file"
+	case tokError:
+		return "error"
+	case tokText:
+		return "text"
+	case tokBraceOpen:
+		return "brace-open"
+	case tokBraceClose:
+		return "brace-close"
+	case tokComma:
+		return "comma"
+	default:
+		return "?"
+	}
+}
+
+type Token struct {
+	Type TokenType
+	Text string
+}
+
+func Scan(text string) (tokens []Token) {
+	var runes []rune
+
+	rr := strings.NewReader(text)
+loop:
+	for {
+		r, _, err := rr.ReadRune()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break loop
+			}
+
+			if len(runes) > 0 {
+				tokens = append(tokens, Token{Text: string(runes), Type: tokText})
+				runes = nil
+			}
```

Comment:
btw. you could extract this as:
```suggestion
	consumeText := func() {
		if len(runes) > 0 {
			tokens = append(tokens, Token{Text: string(runes), Type: tokText})
			runes = nil
		}
	}
```

Reply:
I would like to replace all of that rune writing with taking a slice over the existing string, but yes wrapping it in a function seems nice.

Reply:
Nah, you're both right, appending runes is simpler and probably as efficient, given edge cases of commas, for example.